pkg/trace/api/normalizer.go

```diff
@@ -266,104 +266,110 @@ const maxTagLength = 200
 // backend requirements
-func normalizeTag(tag string) string {
-	if len(tag) == 0 {
+func normalizeTag(v string) string {
+	if len(v) == 0 {
 		return ""
 	}
 	var (
-		trim   int      // start character (if trimming)
-		wiping bool     // true when the previous character has been discarded
-		wipe   [][2]int // sections to discard: (start, end) pairs
-		chars  int      // number of characters processed
+		trim  int      // start character (if trimming)
+		cuts  [][2]int // sections to discard: (start, end) pairs
+		chars int      // number of characters processed
 	)
 	var (
-		i int  // current byte
-		c rune // current rune
+		i    int  // current byte
+		r    rune // current rune
+		jump int  // tracks how much the for range will advance on the next iteration
 	)
-	norm := []byte(tag)
-	for i, c = range tag {
-		if chars >= maxTagLength {
-			// we've reached the maximum
-			break
+	tag := []byte(v)
+	for i, r = range v {
+		jump = utf8.RuneLen(r) // next iteration will be i+jump
+		if r == utf8.RuneError {
+			// on invalid UTF-8, the for range advances only 1 byte;
+			// https://golang.org/ref/spec#For_range (point 2)
+			jump = 1
 		}
 		// fast path; all letters (and colons) are ok
 		switch {
-		case c >= 'a' && c <= 'z' || c == ':':
+		case r >= 'a' && r <= 'z' || r == ':':
 			chars++
-			wiping = false
-			continue
-		case c >= 'A' && c <= 'Z':
+			goto end
+		case r >= 'A' && r <= 'Z':
 			// lower-case
-			norm[i] += 'a' - 'A'
+			tag[i] += 'a' - 'A'
 			chars++
-			wiping = false
-			continue
+			goto end
 		}
 
-		if utf8.ValidRune(c) && unicode.IsUpper(c) {
+		if r != utf8.RuneError && unicode.IsUpper(r) {
 			// lowercase this character
-			if low := unicode.ToLower(c); utf8.RuneLen(c) == utf8.RuneLen(low) {
+			if low := unicode.ToLower(r); utf8.RuneLen(r) == utf8.RuneLen(low) {
 				// but only if the width of the lowercased character is the same;
 				// there are some rare edge-cases where this is not the case, such
 				// as \u017F (ſ)
-				utf8.EncodeRune(norm[i:], low)
-				c = low
+				utf8.EncodeRune(tag[i:], low)
+				r = low
 			}
 		}
 		switch {
-		case unicode.IsLetter(c):
+		case unicode.IsLetter(r):
 			chars++
-			wiping = false
 		case chars == 0:
 			// this character can not start the string, trim
-			trim = i + utf8.RuneLen(c)
-			continue
-		case unicode.IsDigit(c) || c == '.' || c == '/' || c == '-':
+			trim = i + jump
+			goto end
+		case unicode.IsDigit(r) || r == '.' || r == '/' || r == '-':
 			chars++
-			wiping = false
 		default:
 			// illegal character
-			if !wiping {
-				// start a new cut
-				wipe = append(wipe, [2]int{i, i + utf8.RuneLen(c)})
-				wiping = true
+			if n := len(cuts); n > 0 && cuts[n-1][1] >= i {
+				// merge intersecting cuts
+				cuts[n-1][1] += jump
 			} else {
-				// lengthen current cut
-				wipe[len(wipe)-1][1] += utf8.RuneLen(c)
+				// start a new cut
+				cuts = append(cuts, [2]int{i, i + jump})
 			}
 		}
+	end:
+		if i+jump >= 2*maxTagLength {
```

Comment:
Very edge case but shouldn't this be:

```suggestion
		if i+jump >= utf8.UTFMax*maxTagLength {
```

or does the 2 have some other significance here?

Reply:
I think the point here was that if there's a lot of non-digit/non-letter characters to bail early. Added a comment to clarify.